rust/src/mapmanip/procgen/mod.rs

```diff
@@ -0,0 +1,565 @@
+/// that do not meet the above criteria for being part of a "straight" vertical
+/// or horizontal hallway. This submap should be 3x3 tiles, and not anticipate
+/// walls being on any of its sides.
+///
+/// ## JSON Config
+///
+/// A sample mapmanip configuration may look like this:
+///
+/// ```json
+/// {
+///     "type": "MazegenHauberk",
+///     "allowed_area": "/area/mazegen",
+///     "default_floor": "/turf/simulated/floor/plasteel",
+///     "winding_percent": 20,
+///     "extra_connector_chance": 20,
+///     "hallway_horizontal_marker": "/obj/effect/map_effect/marker/mapmanip/submap/insert/deepmaints/hallway_horizontal",
+///     "hallway_vertical_marker": "/obj/effect/map_effect/marker/mapmanip/submap/insert/deepmaints/hallway_vertical",
+///     "hallway_node_marker": "/obj/effect/map_effect/marker/mapmanip/submap/insert/deepmaints/hallway_floor",
+///     "room_configs": [
+///         {
+///             "marker": "/obj/effect/map_effect/marker/mapmanip/submap/insert/deepmaints/small",
+///             "width": 3,
+///             "height": 3,
+///             "min": 7,
+///             "max": 9
+///         },
+///         {
+///             "marker": "/obj/effect/map_effect/marker/mapmanip/submap/insert/deepmaints/medium",
+///             "width": 5,
+///             "height": 5,
+///             "min": 2,
+///             "max": 3
+///         }
+///     ]
+/// }
+/// ```
+///
+/// ## Shortcomings
+///
+/// 1. Currently this mapmanip only works on the 1st z-level of a map file. This
+///    isn't an impossible thing to fix, I just didn't feel like setting it up
+///    to handle other z-levels.
+/// 2. As a result of the mazegen algorithm using odd-sized rooms to ensure
+///    consistent spacing for hallways and junctions, and the fact that the maze
+///    is generated on a grid scaled to 1/3rd the original map size, all rooms
+///    must have sides with dimensions that are both odd and multiples of three,
+///    e.g. 9, 15, 21.
+struct MazegenHauberk {
+    settings: MazegenHauberkSettings,
+    grid: CellGrid,
+    width: i32,
+    height: i32,
+    regions: CellGrid,
+    rooms: Vec<Rect>,
+    current_region: i32,
+}
+
+impl MazegenHauberk {
+    fn new(settings: MazegenHauberkSettings, width: i32, height: i32) -> Self {
+        MazegenHauberk {
+            settings,
+            width,
+            height,
+
+            grid: vec![vec![MAP_TILE_WALL; height as usize]; width as usize],
+            regions: vec![vec![-1; height as usize]; width as usize],
+            rooms: vec![],
+            current_region: -1,
+        }
+    }
+
+    fn can_place_room(&self, room: &Rect) -> bool {
+        if self.rooms.iter().any(|other| room.intersects_with(other)) {
+            return false;
+        }
+
+        for a in room.x1..room.x2 {
+            for b in room.y1..room.y2 {
+                let val = self.grid[a as usize][b as usize];
+                if val == MAP_TILE_IGNORE || val >= 0 {
```

Comment:
Can you add a comment as to what the value being less than 0 implies here? does that mean that it's already occupied?